### PR TITLE
Swim bug fixed 

### PR DESCRIPTION
**Pull request by syed-khadeerahmed**
## Description
Swim module incorrectly upgrades many devices even if one IP is provided

## Type of Change
- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [x] Tasks are idempotent (can be run multiple times without changing state)
- [ ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [x] Playbooks are modular and reusable
- [ ] Handlers are used for actions that need to run on change

## Documentation
- [x] All options and parameters are documented clearly.
- [x] Examples are provided and tested.
- [x] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/swim_workflow_manager.py [hunk 1]
```diff
@@ -810,24 +810,30 @@ def get_device_id(self, params):
...
         """
         device_id = None
-        response = self.dnac._exec(
-            family="devices",
-            function='get_device_list',
-            op_modifies=True,
-            params=params,
-        )
-        self.log("Received API response from 'get_device_list': {0}".format(str(response)), "DEBUG")
+        self.log(params)
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:813]: self.log("Fetching device ID with parameters: {0}".format(params), "INFO")

File: plugins/modules/swim_workflow_manager.py [hunk 2]
```diff
@@ -810,24 +810,30 @@ def get_device_id(self, params):
...
-        device_list = response.get("response")
+            device_list = response.get("response")
 
-        if (len(device_list) == 1):
-            device_id = device_list[0].get("id")
-            self.log("Device Id: {0}".format(str(device_id)), "INFO")
-        else:
-            self.msg = "Device with params: '{0}' not found in Cisco Catalyst Center so can't fetch the device id".format(str(params))
-            self.log(self.msg, "WARNING")
+            if (len(device_list) == 1):
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:825]: if device_list is None:
            self.log("Device list is empty; no devices found for given parameters.", "WARNING")
            raise ValueError("No devices found")

        if len(device_list) == 1:
            device_id = device_list[0].get("id")
            self.log("Successfully retrieved device ID: {0}".format(device_id), "INFO")
            return device_id
        
        self.log("Multiple devices found for parameters: {0}".format(params), "ERROR")
        raise ValueError("Multiple devices found")
        
        
    except ValueError as ve:
        self.msg = "Error: {0}. Unable to fetch unique device ID with parameters: {1}".format(str(ve), params)
        self.status = "failed"
        self.log(self.msg, "ERROR")
        self.result['response'] = self.msg
        return None

    except Exception as e:
        self.msg = "An unexpected error occurred while retrieving device ID: {0}".format(str(e))
        self.status = "failed"
        self.log(self.msg, "ERROR")
        self.result['response'] = self.msg
        return None

File: plugins/modules/swim_workflow_manager.py [hunk 3]
```diff
@@ -810,24 +810,30 @@ def get_device_id(self, params):
             no device or multiple devices match the criteria, it raises an exception.
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:810]: Returns:
        str or None: The unique device ID corresponding to the filtered device, or None if an error occurs.

File: plugins/modules/swim_workflow_manager.py [hunk 4]
```diff
@@ -1703,13 +1715,14 @@ def get_diff_distribution(self):
                         self.result['changed'] = True
                         self.status = "success"
                         self.single_device_distribution = True
-                        self.result['msg'] = "Image with Id {0} Distributed Successfully".format(image_id)
-                        self.result['response'] = self.msg
+                        self.result['msg'] = "Image with Id {0} Distributed Successfully for the device: {1} ".format(image_id, device_ip)
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1718]: self.result['msg'] = "Image '{0}' with ID {1} has been successfully distributed to device: {2}".format(image_name, image_id, device_ip)
OR
self.result['msg'] = "Successfully distributed image '{0}' (ID: {1}) to device: {2}".format(image_name, image_id, device_ip)

File: plugins/modules/swim_workflow_manager.py [hunk 5]
```diff
@@ -1703,13 +1715,14 @@ def get_diff_distribution(self):
...
+                        self.result['response'] = self.result['msg']
+                        self.log(self.result['msg'])
                         break
 
                     if task_details.get("isError"):
                         self.status = "failed"
-                        self.msg = "Image with Id {0} Distribution Failed".format(image_id)
+                        self.msg = "Image with Id {0} Distribution Failed for the device : {1}".format(image_id, device_ip)
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1725]: Same as above

File: plugins/modules/swim_workflow_manager.py [hunk 6]
```diff
@@ -1730,6 +1743,7 @@ def get_diff_distribution(self):
         distribution_task_dict = {}
 
         for device_uuid in device_uuid_list:
+            self.log("Stating image distribution for multiple devices")
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1746]: what image?

File: plugins/modules/swim_workflow_manager.py [hunk 7]
```diff
@@ -1673,8 +1683,10 @@ def get_diff_distribution(self):
         self.complete_successful_distribution = False
         self.partial_successful_distribution = False
         self.single_device_distribution = False
+        device_ip = self.want.get("distribution_details").get("device_ip_address")
 
         if self.have.get("distribution_device_id"):
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1688]: Print image version or image name ... 
```
distribution_device_id = self.have.get("distribution_device_id")
if distribution_device_id:
    self.log("Starting image distribution for device IP {0} with ID {1}, targeting software version {2}.".format(
        device_ip, distribution_device_id, image_name
    ), "INFO")
```

File: plugins/modules/swim_workflow_manager.py [hunk 8]
```diff
@@ -1799,8 +1813,10 @@ def get_diff_activation(self):
         self.complete_successful_activation = False
         self.partial_successful_activation = False
         self.single_device_activation = False
+        device_ip = self.want.get("activation_details").get("device_ip_address")
 
         if self.have.get("activation_device_id"):
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1818]: Same as like getting distribution_device_id...

File: plugins/modules/swim_workflow_manager.py [hunk 9]
```diff
@@ -1861,6 +1877,7 @@ def get_diff_activation(self):
         activation_task_dict = {}
 
         for device_uuid in device_uuid_list:
+            self.log("Stating image activation for multiple devices")
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1880]: what image ?? Check the above log messages and make more meaningful

File: plugins/modules/swim_workflow_manager.py (hunk 3 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:810]: Updated the code

File: plugins/modules/swim_workflow_manager.py (hunk 1 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:813]: updated the log

File: plugins/modules/swim_workflow_manager.py (hunk 7 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:1688]: updated the code

File: plugins/modules/swim_workflow_manager.py (hunk 2 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:825]: updated the code

File: plugins/modules/swim_workflow_manager.py (hunk 4 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:1718]: updated the code

File: plugins/modules/swim_workflow_manager.py (hunk 8 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:1872]: updated the code

File: plugins/modules/swim_workflow_manager.py (hunk 9 above)
Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py:1937]: updated the code

File: plugins/modules/swim_workflow_manager.py [hunk 10]
```diff
@@ -1144,10 +1159,30 @@ def get_have(self):
                 managementIpAddress=distribution_details.get("device_ip_address"),
                 macAddress=distribution_details.get("device_mac_address"),
             )
+
             device_id = self.get_device_id(device_params)
 
-            if device_id is not None:
-                have["distribution_device_id"] = device_id
+            if (distribution_details.get("device_ip_address") or
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:1165]: I don't understand why do we need to get device_ip/mac... again? We already got those values in 1156?
Can we replace the whole logic from 1156?
Also why do we use camel case in 1156? Can we replace?

```
device_params = {
    "hostname": distribution_details.get("device_hostname"),
    "serial_number": distribution_details.get("device_serial_number"),
    "management_ip_address": distribution_details.get("device_ip_address"),
    "mac_address": distribution_details.get("device_mac_address")
}

if any(device_params.values()):
    device_id = self.get_device_id(device_params)

    if device_id is None:
        # Device not found, format error message
        params_list = []
        for key, value in device_params.items():
            if value:
                formatted_param = "{0}: {1}".format(key, value)
                params_list.append(formatted_param)

        # Join the parameters for the final error message
        params_message = ", ".join(params_list)
        self.status = "failed"
        self.msg = "The device with the following parameter(s): {0} could not be found in the Cisco Catalyst Center.".format(params_message)
        self.log(self.msg, "ERROR")
        self.result['response'] = self.msg
        self.check_return_status()
    else:
        # Device found, log success and store the device ID
        self.log("Device with ID {0} found and added to distribution details.".format(device_id), "DEBUG")
        have["distribution_device_id"] = device_id
else:
    self.log("No valid device identifier found in the distribution details.", "WARNING")
    What we do here ....???? 
```

Comment (syed-khadeerahmed) [reply plugins/modules/swim_workflow_manager.py]: updated the code

File: plugins/modules/swim_workflow_manager.py [hunk 11]
```diff
@@ -256,8 +256,8 @@
             type: str
       image_activation_details:
         description: Details for SWIM image activation. Device on which the image needs to activated
```
Comment (madhansansel) [inline plugins/modules/swim_workflow_manager.py:258]: Same as like distribution

**Review by madhansansel — APPROVED** [review]

**Pull request merged**